File: Add Code Here/C++/dfs.cpp
#include<bits/stdc++.h>
using namespace std;
// C++ code to implement DFS 

class Graph{

    int v;
    vector<vector<int>> adj;
    vector<bool> visited;
public:
    Graph(int n){
        this->v = n;
        adj.resize(v);
        visited.resize(v);
    }

    void AddEdge(int a, int b, bool undir){
        adj[a].push_back(b);
        if (undir)
        {
            adj[b].push_back(a);
        }
    }

    void PrintGraph(){
        for (int i = 0; i < adj.size(); i++)
        {
            cout<<i<<"-->";
            for(auto j: adj[i]){
                cout<<j<<" ,";
            }
            cout<<endl;
        }   
    }

    bool allVisited(vector<bool> a){
        int sz = a.size();

        for(auto i: a){
            if (i == false)
            {
                return false;
            }
            
        }

        return true;
    }

    void dfs(int source){  // recursive traversal

        if (allVisited(visited) == true)
        {
            return;
        }
        
        cout<<source<<" ";
        visited[source] = true;
        for (int i = 0; i < adj[source].size(); i++)
        {
            if (visited[adj[source][i]] == false)
            {
                dfs(adj[source][i]);
            }
        }
        return;
    }
    
};

int main(){
    
    Graph g(7);
    g.AddEdge(1 ,0 ,1);
    g.AddEdge(1 ,2 ,1);
    g.AddEdge(0 ,4 ,1);
    g.AddEdge(2 ,3 ,1);
    g.AddEdge(3 ,4 ,1);
    g.AddEdge(3 ,5 ,1);
    g.AddEdge(4 ,5 ,1);
    g.AddEdge(5 ,6 ,1);

    g.dfs(1);
    
    return 0;
}

File: meister2/dfs.py
# An Iterative Python program to do DFS traversal from
# a given source vertex. DFS(int s) traverses vertices
# reachable from s.

# This class represents a directed graph using adjacency
# list representation
class Graph:
	def __init__(self,V): # Constructor
		self.V = V	 # No. of vertices
		self.adj = [[] for i in range(V)] # adjacency lists

	def addEdge(self,v, w):	 # to add an edge to graph
		self.adj[v].append(w) # Add w to v’s list.


	# prints all not yet visited vertices reachable from s
	def DFS(self,s):		 # prints all vertices in DFS manner from a given source.
								# Initially mark all vertices as not visited
		visited = [False for i in range(self.V)]

		# Create a stack for DFS
		stack = []

		# Push the current source node.
		stack.append(s)

		while (len(stack)):
			# Pop a vertex from stack and print it
			s = stack[-1]
			stack.pop()

			# Stack may contain same vertex twice. So
			# we need to print the popped item only
			# if it is not visited.
			if (not visited[s]):
				print(s,end=' ')
				visited[s] = True

			# Get all adjacent vertices of the popped vertex s
			# If a adjacent has not been visited, then push it
			# to the stack.
			for node in self.adj[s]:
				if (not visited[node]):
					stack.append(node)



# Driver program to test methods of graph class

g = Graph(5); # Total 5 vertices in graph
g.addEdge(1, 0);
g.addEdge(0, 2);
g.addEdge(2, 1);
g.addEdge(0, 3);
g.addEdge(1, 4);

print("Following is Depth First Traversal")
g.DFS(0)

# This code is contributed by ankush_953
